File: pump/equil/src/main/java/app/aaps/pump/equil/manager/command/BaseSetting.java
package app.aaps.pump.equil.manager.command;

import java.nio.ByteBuffer;

import app.aaps.core.interfaces.logging.LTag;
import app.aaps.pump.equil.manager.AESUtil;
import app.aaps.pump.equil.manager.EquilCmdModel;
import app.aaps.pump.equil.manager.EquilResponse;
import app.aaps.pump.equil.manager.Utils;


public abstract class BaseSetting extends BaseCmd {

    public BaseSetting(long createTime) {
        super(createTime);
    }

    public byte[] getReqData() {
        byte[] indexByte = Utils.intToBytes(pumpReqIndex);
        byte[] tzm = Utils.hexStringToBytes(getEquilDevices());
        byte[] data = Utils.concat(indexByte, tzm);
        pumpReqIndex++;
        return data;
    }


    @Override
    public EquilResponse getEquilResponse() {
        config = false;
        isEnd = false;
        response = new EquilResponse(createTime);
        byte[] pwd = Utils.hexStringToBytes(getEquilPassWord());
        byte[] data = getReqData();
        EquilCmdModel equilCmdModel;
        try {
            equilCmdModel = AESUtil.aesEncrypt(pwd, data);
            return responseCmd(equilCmdModel, defaultPort + "0000");
        } catch (Exception e) {
            synchronized (this) {
                setCmdStatus(false);
            }
        }
        return null;
    }


    public EquilResponse decodeEquilPacket(byte[] data) {
        if (!checkData(data)) {
            return null;
        }
        byte code = data[4];
        int intValue = getIndex(code);
        if (config) {
            if (rspIndex == intValue) {
                return null;
            }
            boolean flag = isEnd(code);
            ByteBuffer buffer = ByteBuffer.wrap(data);
            response.add(buffer);
            if (!flag) {
                return null;
            }
            try {
                EquilResponse response1 = decodeConfirm();
                isEnd = true;
                response = new EquilResponse(createTime);
                rspIndex = intValue;
                return response1;
            } catch (Exception e) {
                response = new EquilResponse(createTime);
                aapsLogger.debug(LTag.PUMPCOMM, "decodeEquilPacket error =====" + e.getMessage());
            }
            return null;
        }
        boolean flag = isEnd(code);
        ByteBuffer buffer = ByteBuffer.wrap(data);
        response.add(buffer);
        if (!flag) {
            return null;
        }
        try {
            EquilResponse list = decode();
            response = new EquilResponse(createTime);
            config = true;
            rspIndex = intValue;
            return list;
        } catch (Exception e) {
            response = new EquilResponse(createTime);
            aapsLogger.debug(LTag.PUMPCOMM, "decodeEquilPacket error=====" + e.getMessage());
        }
        return null;

    }

    @Override
    public EquilResponse decode() throws Exception {
        EquilCmdModel reqModel = decodeModel();
        byte[] pwd = Utils.hexStringToBytes(getEquilPassWord());
        String content = AESUtil.decrypt(reqModel, pwd);
        String pwd2 = content.substring(8);
        runPwd = pwd2;
        byte[] data = getFirstData();
        EquilCmdModel equilCmdModel = AESUtil.aesEncrypt(Utils.hexStringToBytes(runPwd), data);
        runCode = reqModel.getCode();
        return responseCmd(equilCmdModel, port + runCode);
    }

    public EquilResponse getNextEquilResponse() {
        aapsLogger.debug(LTag.PUMPCOMM, "getNextEquilResponse=== start11 ");
        config = true;
        isEnd = false;
        response = new EquilResponse(createTime);
        byte[] data = getFirstData();
        EquilCmdModel equilCmdModel;
        try {
            aapsLogger.debug(LTag.PUMPCOMM, "getNextEquilResponse=== start ");
            equilCmdModel = AESUtil.aesEncrypt(Utils.hexStringToBytes(runPwd), data);
            return responseCmd(equilCmdModel, port + runCode);
        } catch (Exception e) {
            aapsLogger.debug(LTag.PUMPCOMM, "getNextEquilResponse===" + e.getMessage());
            synchronized (this) {
                setCmdStatus(false);
            }
        }
        return null;
    }


    public abstract byte[] getFirstData();

    public abstract byte[] getNextData();

    public abstract void decodeConfirmData(byte[] data);

    public EquilResponse decodeConfirm() throws Exception {
        EquilCmdModel equilCmdModel = decodeModel();
        runCode = equilCmdModel.getCode();
        String content = AESUtil.decrypt(equilCmdModel, Utils.hexStringToBytes(runPwd));
        decodeConfirmData(Utils.hexStringToBytes(content));
        byte[] data = getNextData();
        EquilCmdModel equilCmdModel2 = AESUtil.aesEncrypt(Utils.hexStringToBytes(runPwd), data);
        return responseCmd(equilCmdModel2, port + runCode);
    }
}

File: pump/equil/src/main/kotlin/app/aaps/pump/equil/manager/command/BaseSetting.kt
package app.aaps.pump.equil.manager.command

import app.aaps.core.interfaces.logging.AAPSLogger
import app.aaps.core.interfaces.logging.LTag
import app.aaps.core.keys.interfaces.Preferences
import app.aaps.pump.equil.manager.AESUtil
import app.aaps.pump.equil.manager.EquilCmdModel
import app.aaps.pump.equil.manager.EquilManager
import app.aaps.pump.equil.manager.EquilResponse
import app.aaps.pump.equil.manager.Utils
import java.lang.Exception
import java.nio.ByteBuffer

abstract class BaseSetting(
    createTime: Long,
    aapsLogger: AAPSLogger,
    preferences: Preferences,
    equilManager: EquilManager
) : BaseCmd(createTime, aapsLogger, preferences, equilManager) {

    fun getReqData(): ByteArray {
        val indexByte = Utils.intToBytes(pumpReqIndex)
        val tzm = Utils.hexStringToBytes(getEquilDevices())
        val data = Utils.concat(indexByte, tzm)
        pumpReqIndex++
        return data
    }

    override fun getEquilResponse(): EquilResponse? {
        config = false
        isEnd = false
        response = EquilResponse(createTime)
        val pwd = Utils.hexStringToBytes(getEquilPassWord())
        val data = getReqData()
        var equilCmdModel: EquilCmdModel?
        try {
            equilCmdModel = AESUtil.aesEncrypt(pwd, data)
            return responseCmd(equilCmdModel, DEFAULT_PORT + "0000")
        } catch (_: Exception) {
            synchronized(this) {
                cmdStatus = false
            }
        }
        return null
    }

    override fun decodeEquilPacket(data: ByteArray): EquilResponse? {
        if (!checkData(data)) {
            return null
        }
        val code = data[4]
        val intValue = getIndex(code)
        if (config) {
            if (rspIndex == intValue) {
                return null
            }
            val flag = isEnd(code)
            val buffer = ByteBuffer.wrap(data)
            response!!.add(buffer)
            if (!flag) {
                return null
            }
            try {
                val response1 = decodeConfirm()
                isEnd = true
                response = EquilResponse(createTime)
                rspIndex = intValue
                return response1
            } catch (e: Exception) {
                response = EquilResponse(createTime)
                aapsLogger.debug(LTag.PUMPCOMM, "decodeEquilPacket error =====" + e.message)
            }
            return null
        }
        val flag = isEnd(code)
        val buffer = ByteBuffer.wrap(data)
        response!!.add(buffer)
        if (!flag) {
            return null
        }
        try {
            val list = decode()
            response = EquilResponse(createTime)
            config = true
            rspIndex = intValue
            return list
        } catch (e: Exception) {
            response = EquilResponse(createTime)
            aapsLogger.debug(LTag.PUMPCOMM, "decodeEquilPacket error=====" + e.message)
        }
        return null
    }

    override fun decode(): EquilResponse? {
        val reqModel = decodeModel()
        val pwd = Utils.hexStringToBytes(getEquilPassWord())
        val content = AESUtil.decrypt(reqModel, pwd)
        val pwd2 = content.substring(8)
        runPwd = pwd2
        val data = getFirstData()
        val equilCmdModel = AESUtil.aesEncrypt(Utils.hexStringToBytes(pwd2), data)
        runCode = reqModel.code
        return responseCmd(equilCmdModel, port + runCode)
    }

    override fun getNextEquilResponse(): EquilResponse? {
        aapsLogger.debug(LTag.PUMPCOMM, "getNextEquilResponse=== start11 ")
        config = true
        isEnd = false
        response = EquilResponse(createTime)
        val data = getFirstData()
        try {
            aapsLogger.debug(LTag.PUMPCOMM, "getNextEquilResponse=== start ")
            var equilCmdModel: EquilCmdModel = AESUtil.aesEncrypt(Utils.hexStringToBytes(runPwd!!), data)
            return responseCmd(equilCmdModel, port + runCode)
        } catch (e: Exception) {
            aapsLogger.debug(LTag.PUMPCOMM, "getNextEquilResponse===" + e.message)
            synchronized(this) {
                cmdStatus = false
            }
        }
        return null
    }

    abstract fun getFirstData(): ByteArray?

    abstract fun getNextData(): ByteArray?

    abstract fun decodeConfirmData(data: ByteArray)

    override fun decodeConfirm(): EquilResponse? {
        val equilCmdModel = decodeModel()
        runCode = equilCmdModel.code
        val content = AESUtil.decrypt(equilCmdModel, Utils.hexStringToBytes(runPwd!!))
        decodeConfirmData(Utils.hexStringToBytes(content))
        val data = getNextData()
        val equilCmdModel2 = AESUtil.aesEncrypt(Utils.hexStringToBytes(runPwd!!), data)
        return responseCmd(equilCmdModel2, port + runCode)
    }
}
